[FLINK-21351][checkpointing] Don't subsume last checkpoint

rkhachatryan opened the pull request:
## What is the purpose of the change

 ```
When a savepoint is added to CompletedCheckpointStore
all previous checkpoints will be removed if number to
retain is 1.

This makes future incremental checkpoints invalid since
they can refer to the discarded state.
```

## Verifying this change

Added unit tests for the `DefaultCompletedCheckpointStore` (`testAtLeastOneCheckpointRetained`, `testOlderSavepointSubsumed`, `testSubsumeAfterStoppingWithSavepoint`, `testNotSubsumedIfNotNeeded`).

I test `DefaultCompletedCheckpointStore` and not the `CheckpointSubsumeHelper` to verify their integration.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

pnowojski reviewed (changes requested):
Thanks for the bug fix @rkhachatryan. I've left a couple of comments.

Also I'm not sure if this approach is worth the extra complexity ( https://issues.apache.org/jira/browse/FLINK-21351?focusedCommentId=17286454&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17286454 ).

What's the problem with the original suggestion to just ignore savepoints? That the simpler solution would keep savepoints forever and never discard/subsume them?

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/StandaloneCompletedCheckpointStoreTest.java:
```diff
@@ -89,37 +82,6 @@ public void testSuspendDiscardsCheckpoints() throws Exception {
         verifyCheckpointDiscarded(taskStates);
     }
 
-    /**
-     * Tests that the checkpoint does not exist in the store when we fail to add it into the store
-     * (i.e., there exists an exception thrown by the method).
-     */
-    @Test
-    public void testAddCheckpointWithFailedRemove() throws Exception {
-
-        final int numCheckpointsToRetain = 1;
-        CompletedCheckpointStore store =
-                createCompletedCheckpoints(numCheckpointsToRetain, Executors.directExecutor());
-
-        for (long i = 0; i <= numCheckpointsToRetain; ++i) {
-            CompletedCheckpoint checkpointToAdd = mock(CompletedCheckpoint.class);
-            doReturn(i).when(checkpointToAdd).getCheckpointID();
-            doReturn(Collections.emptyMap()).when(checkpointToAdd).getOperatorStates();
-            doThrow(new IOException()).when(checkpointToAdd).discardOnSubsume();
-
-            try {
-                store.addCheckpoint(checkpointToAdd, new CheckpointsCleaner(), () -> {});
-
-                // The checkpoint should be in the store if we successfully add it into the store.
-                List<CompletedCheckpoint> addedCheckpoints = store.getAllCheckpoints();
-                assertTrue(addedCheckpoints.contains(checkpointToAdd));
-            } catch (Exception e) {
-                // The checkpoint should not be in the store if any exception is thrown.
-                List<CompletedCheckpoint> addedCheckpoints = store.getAllCheckpoints();
-                assertFalse(addedCheckpoints.contains(checkpointToAdd));
-            }
-        }
-    }
-
```
Thanks for pointing out, you're right.
I'll remove the dead code and replace mocks with the normal objects.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointSubsumeHelper.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.util.Deque;
+
+class CheckpointSubsumeHelper {
```
Good point, I'll add the docs for this class.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointSubsumeHelper.java:
```diff
@@ -0,0 +1,47 @@
+    public static void subsume(
+            Deque<CompletedCheckpoint> checkpoints,
+            int numRetain,
+            ThrowingConsumer<CompletedCheckpoint, Exception> subsumeAction)
```
To me, all those approaches are not ideal:
1. Function (`ThrowingConsumer`) - harder to find implementations
1. Dedicated interface - not clear what methods it has just from looking at it as an argument
1. Returning a collection - a) duplicated error handling b) non-obvious contract that the original queue is modified too (or needs to be replaced); 

I think I'll go with the 2nd approach (though I personally would still prefer 1st =) ).

pnowojski replied:
Maybe one last try from my side to avoid functional interfaces here

> non-obvious contract that the original queue is modified too (or needs to be replaced);

With the method name like `findAndRemoveCheckpointsToSubsume()` or something like that, it would be relatively obvious. Also in your current code it's (maybe even more) unclear that this `subsume()` is modifying `Deque<CompletedCheckpoint> checkpoints`.

> a) duplicated error handling

That would be minuscule duplication (4 lines of code in two 5 LOC methods).


For me 1 <<< 2 < 3. If you still prefer 2 over 3, let's go with that. 

The difference between 2 and 3 is not that large, but I'm afraid in the future I would be regretting it :( Similarly how I didn't like 1. in other places that we introduced them in the last year or so, but once I had to debug/analyse this newly added code a couple of times I changed my mind from "slightly against" to "strongly against". If you want, let's see, maybe 2. will be good enough.

rkhachatryan replied:
Good, I'll keep a dedicated interface.

rkhachatryan commented:
Thanks for the review @pnowojski, I've addressed all your comments.

> What's the problem with the original suggestion to just ignore savepoints? 

My main concern was that subtracting savepoint count from total count would keep savepoints in the queue forever.
Besides that, we'd have to maintain a seperatae counter. We'd have to update it in all tree implementations. We'd have to count savepoints on restore. So I don't think it would be much more simple in the end.

I also considered a third option to not enter the loop at all if we are adding a savepoint. That would be indeed simple, but:
- with checkpointing disabled this could lead to the same problem (keeping savepoints forever)
- the invariant wouldn't be clear (as we discussed offline)

So all in all I believe the current approach is preferable.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointSubsumeHelper.java:
```diff
@@ -87,6 +87,8 @@ private static boolean canSubsume(
             Optional<CompletedCheckpoint> latestNonSavepoint) {
         if (next == latest) {
             return false;
+        } else if (next.getProperties().isSavepoint()) {
+            return true;
```
Add a unit test for this case?

rkhachatryan replied:
This is already covered by `CheckpointCoordinatorTest.testTriggerAndConfirmSimpleSavepoint`.

pnowojski reviewed (approved):
Thanks for the fix @rkhachatryan . LGTM (Assuming build is/will be green)

rkhachatryan commented:
Thanks for the review. Merging.
